solo-chains/node/tanssi-relay-service/src/lib.rs

```diff
@@ -14,6 +14,602 @@
+            client: client.clone(),
+            transaction_pool: transaction_pool.clone(),
+            spawn_handle: task_manager.spawn_handle(),
+            import_queue,
+            block_announce_validator_builder: None,
+            warp_sync_params: None,
+            block_relay: None,
+            metrics,
+        })?;
+
+    if config.offchain_worker.enabled {
+        use futures::FutureExt;
+
+        task_manager.spawn_handle().spawn(
+            "offchain-workers-runner",
+            "offchain-work",
+            sc_offchain::OffchainWorkers::new(sc_offchain::OffchainWorkerOptions {
+                runtime_api_provider: client.clone(),
+                keystore: Some(keystore_container.keystore()),
+                offchain_db: backend.offchain_storage(),
+                transaction_pool: Some(OffchainTransactionPoolFactory::new(
+                    transaction_pool.clone(),
+                )),
+                network_provider: Arc::new(network.clone()),
+                is_validator: role.is_authority(),
+                enable_http_requests: false,
+                custom_extensions: move |_| vec![],
+            })
+            .run(client.clone(), task_manager.spawn_handle())
+            .boxed(),
+        );
+    }
+
+    let mut command_sink = None;
+
+    if role.is_authority() {
+        let proposer = sc_basic_authorship::ProposerFactory::new(
+            task_manager.spawn_handle(),
+            client.clone(),
+            transaction_pool.clone(),
+            prometheus_registry.as_ref(),
+            telemetry.as_ref().map(|x| x.handle()),
+        );
+
+        let commands_stream: Box<
+            dyn Stream<Item = EngineCommand<<Block as BlockT>::Hash>> + Send + Sync + Unpin,
+        > = match sealing {
+            Sealing::Instant => {
+                Box::new(
+                    // This bit cribbed from the implementation of instant seal.
+                    transaction_pool
+                        .pool()
+                        .validated_pool()
+                        .import_notification_stream()
+                        .map(|_| EngineCommand::SealNewBlock {
+                            create_empty: false,
+                            finalize: false,
+                            parent_hash: None,
+                            sender: None,
+                        }),
+                )
+            }
+            Sealing::Manual => {
+                let (sink, stream) = futures::channel::mpsc::channel(1000);
+                // We wait for slot duration here to make sure, we have strictly increasing slot numbers
+                // otherwise babe block import will reject the block.
+                // There might be more efficient way of doing this with stream throttling, but
+                // for development purposes it is okay.
+                let wrapped_stream = stream.then(move |item| {
+                    futures_time::task::sleep(slot_duration.as_duration().into())
```

Comment:
It's curious that this is not needed for aura, I guess because it doesn't enforce increasing slot numbers?

Reply:
Yes. Babe block import enforces this.

Reply:
I just remembered that we mock the timestamp for aura:

https://github.com/moondance-labs/tanssi/blob/dfb7b168dc20aae0690732d7fdd8f1a5b246dff6/node/src/service.rs#L107

So maybe you can do something similar here?

Reply:
Yes let's do something similar

Reply:
Yes. I am now utilizing aux storage to store last timestamp and at minimum incrementing by `slot_duration`. 

Also added a typescript test to create bunch of blocks to test it and it seems to work fine.

Reply:
alright so is essentially the same but instead of using a local thread we use an storage item right? I dont see any caveats  doing that right @tmpolaczyk ?

Reply:
yes.

Reply:
Actually I prefer using storage items because they handle rollbacks automatically, but I guess thread_locals are cheaper to use. Either way is fine for tests.